Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs

```diff
@@ -387,7 +387,7 @@ private bool ShouldReconnect
             }
 
             // We don't know what happened... try reconnecting?
-            return true;
+            return iterationResult.Inner == null || ShouldReconnect(iterationResult.Inner, out shouldTerminate, out withNewSession);
```

Comment:
Instead of recursively calling this method, use `.Unwrap()` on the input parameter to access the innermost error.